### ISSUE TITLE
Non-ASCII clipboard broken with Wayland applications

### ISSUE THREAD

**Issue by CendioOssman** [fltk/fltk#182]
We've noticed that the clipboard is broken with modern Wayland applications (might just be GTK+ applications). Everything non-ASCII gets replaces with a `\uxxxx` string instead.

After some digging this is caused by this incorrect code in FLTK:

https://github.com/fltk/fltk/blob/d7af409f426fba1790149499dd26614b137a771a/src/Fl_x.cxx#L1471

`text/plain` is US-ASCII, so it should not be among the preferred format. It used to work by pure luck that `text/plain;charset=utf-8` was listed before `text/plain` in most cases. Somewhere along the line the order was changed though and exposed this bug.

This patch is tested on FLTK 1.3.5, but the code seems to be the same on `master`:

```diff
diff -up fltk-1.3.5/src/Fl_x.cxx.text fltk-1.3.5/src/Fl_x.cxx
--- fltk-1.3.5/src/Fl_x.cxx.text	2021-01-20 14:10:21.097838241 +0100
+++ fltk-1.3.5/src/Fl_x.cxx	2021-01-20 14:11:16.481100168 +0100
@@ -1425,13 +1425,13 @@ fprintf(stderr,"\n");*/
 	  t = ((Atom*)portion)[i];
 	  if (t == fl_Xatextplainutf ||
 	      t == fl_Xatextplainutf2 ||
-	      t == fl_Xatextplain ||
 	      t == fl_XaUtf8String) {
 	    type = t;
 	    break;
 	  }
 	  // rest are only used if no utf-8 available:
 	  if (t == fl_XaText ||
+	      t == fl_Xatextplain ||
 	      t == fl_XaTextUriList ||
 	      t == fl_XaCompoundText) type = t;
 	}
@@ -1671,13 +1671,13 @@ fprintf(stderr,"\n");*/
         //printf("fl_dnd_source_types[%d]=%ld(%s)\n",i,t,XGetAtomName(fl_display,t));
         if (t == fl_Xatextplainutf ||		// "text/plain;charset=UTF-8"
             t == fl_Xatextplainutf2 ||		// "text/plain;charset=utf-8" -- See STR#2930
-            t == fl_Xatextplain ||		// "text/plain"
             t == fl_XaUtf8String) {		// "UTF8_STRING"
           type = t;
           break;
 	}
         // rest are only used if no utf-8 available:
         if (t == fl_XaText ||			// "TEXT"
+            t == fl_Xatextplain ||		// "text/plain"
             t == fl_XaTextUriList ||		// "text/uri-list"
             t == fl_XaCompoundText) type = t;	// "COMPOUND_TEXT"
       }
```

Downstream issue:

https://www.cendio.com/bugzilla/show_bug.cgi?id=7627

**Comment by Albrecht-S** [comment]
PS, FWIW (testing, internal info): the original bug report (aka "Downstream issue") mentions:
> ... X11 clipboard, checked using "xclip -o -selection clipboard -target TARGETS" ...

which appears to be very helpful for testing. Thanks for this as well.

**Comment by CendioOssman** [comment]
> Is this true?

Correct. Seen with every Wayland application I tried. But they are all GTK+, so it might just be that toolkit that provokes this.

> If yes, does this also happen with drag and drop?
> 

Didn't originally test that, but just did and it seems to work fine.

> Unfortunately I don't have a wayland system I can test with. Is there another test case you're aware of that would show a similar/equal result? Can you describe such a scenario for easy testing?

We haven't seen this anywhere else in the real world. But you should be able to provoke it with a custom build of FLTK that lists "text/plain" first in TARGETS.

(Wayland also works with in a virtual machine, if that is an option for you)

**Comment by Albrecht-S** [comment]
@CendioOssman Thanks, yes, a VM is fine. I have one with Debian (not tested yet) and I can install Ubuntu in another one. Can you give me some hints as to what I need to do to:
- would bare Ubuntu 20.04 (LTS) be OK? I see [in your report](https://www.cendio.com/bugzilla/show_bug.cgi?id=7627) "_Ubuntu 20.04, gnome 3.36.3, Broken_" so I guess it would do it
- what do I need to do to install/activate/run Wayland? Is it the default or ... ? [1]
- can you recommend a specific "Wayland application" that shows the issue?

BTW, I tested on my Linux Mint with "Gimp" (which I assume ;-) to be a "GTK application") and it didn't exhibit the issue.

I'll go ahead and test some other things (as you recommended) but I'd appreciate if you could help me with some of the questions above. TIA.

[1] ISTR that Wayland was the default in a certain Ubuntu version but they switched back, but that's only faint memory.

**Comment by CendioOssman** [comment]
> * would bare Ubuntu 20.04 (LTS) be OK? I see [in your report](https://www.cendio.com/bugzilla/show_bug.cgi?id=7627) "_Ubuntu 20.04, gnome 3.36.3, Broken_" so I guess it would do it
> 

Yes. As you spotted that's what we used to test here.

> * what do I need to do to install/activate/run Wayland? Is it the default or ... ? [1]
> 

Not sure if it's the default or not. But to make sure, click the username during login but don't enter a password. In the corner theres a cog icon where you can select the desktop environment. Pick the one that just says "GNOME", not the one that says "GNOME on Xorg".

> * can you recommend a specific "Wayland application" that shows the issue?

I tested with gedit and gnome-terminal.

**Comment by CendioOssman** [comment]
And to verify, you can run the commands "xprop" or "xwininfo". They will only show their target cursor over X11 windows, not Wayland ones.

**Comment by Albrecht-S** [comment]
@CendioOssman Thanks for the info, I'll try that.

**Comment by Albrecht-S** [comment]
I made same progress: my Ubuntu 20.04 (LTS) VM is up and running although not yet configured to build FLTK. I could log in to the _wayland desktop_ and I can confirm the differently ordered clipboard targets that lead to the observed result (using `xclip` in my first tests).

@CendioOssman: I am pretty sure that your analysis is correct and I could also see the unexpected behavior with FLTK's test/editor -- built on my main system and copied over to the Ubuntu system for my first tests. But I also found other critical issues which I'm still investigating (one of these seems to be a {wayland | system | whatever} bug which I'll investigate further). I'll let you know what I will find out.

Regarding your patch: yes, I think it would work, but I believe the algorithm we are using now is "_suboptimal_" anyway. I intend to rewrite it for better (more reliable) results to find the **_best_ matching target** and not **_any_ matching target** (as it is now, basically). There are some issues with the current approach...

**Comment by Albrecht-S** [comment]
@CendioOssman: I can now offer a patch '[issue-182_1.3_v1.diff.txt](https://github.com/fltk/fltk/files/5890600/issue-182_1.3_v1.diff.txt)' that seems to work for me for FLTK branch-1.3. It's much longer than anticipated and fixes multiple issues:
1. copy and paste error, as described in this issue
2. same issue for drag and drop
3. suboptimal target selection in other cases
4. code duplication.

The fix for (3) includes an algorithm that selects the "best" target from a list provided internally by the FLTK lib. This list is ordered by preference such that we should always **select the best target** _from all available targets_. The old code would select one *possible* target and not the *best* target (even with your proposed fix). One of the functions implementing this algorithm (find_target_text() or find_target_image()) is now always called (4) to select the target to be requested from the source program (copy/paste or dnd, respectively).

In my tests on my original Linux Mint system and on my Ubuntu/Wayland system with both types of programs on the latter seemed to work fine.

I'd appreciate if you could test the patch and give your feedback. TIA.

PS: I'll double check the changes later and I'll try to find out if I missed anything but your feedback would be welcome.
PPS: I prepared the same changes for 1.4 as well, but none of them is already committed.

**Comment by Albrecht-S** [comment]
Side note: the bug I mentioned before can be seen as follows:
- copy something in a "Wayland editor" (`gedit` works) or "Wayland filemanager"
- use `xclip -o -selection clipboard -target TARGETS`
- subsequently you may use all available targets with `xclip -o -selection clipboard -target "target-n"` with "target-n" from the list of targets (note: quoting is necessary for some targets)
- expect some of the targets to fail (error message from `xclip`, this seems to be "normal", e.g. `Error: target MULTIPLE not available`)
- in some cases `xclip` will "hang" but can still be stopped with `ctrl/c`
- in those cases mentioned above the source application may "hang" infinitely
- if that happens you may see a system error message like this: ![gedit-hangs.png](https://user-images.githubusercontent.com/26235500/106214829-ec914500-61cf-11eb-8dca-709ece572965.png)
- you must stop the application with the "Force Quit" button or some other means (e.g. `killall gedit`)

In the case shown above `top` shows `gedit` with more than 100% CPU load and `xclip` failed with target type "TEXT" whereas "text/plain", "text/plain;charset=utf-8", and "STRING" worked fine.

@CendioOssman and others: I'm not sure if I can or if it's worth to file a bug report (where?) but before I think further about it, can you confirm this issue?

**Comment by Albrecht-S** [comment]
@CendioOssman: Hi Pierre, are there any news about my patch ['issue-182_1.3_v1.diff.txt'](https://github.com/fltk/fltk/files/5890600/issue-182_1.3_v1.diff.txt)? Could you find time to test it? Does it work solve the issue for you?

I'd like to commit/push it but I'd appreciate to see a confirmation before I do this. TIA.

**Comment by CendioOssman** [comment]
Sorry, things are a bit hectic here at the moment. I'll see if I can get it tested during the week.

**Comment by Albrecht-S** [comment]
Thanks, I'll wait until I hear (read) from you.

**Comment by CendioOssman** [comment]
I can confirm that the patch works with Ubuntu 20.04 and GNOME. Both the FLTK included in the system, and the older one here at Cendio saw the bug. With your patch the bug goes away.

I had to massage the patch a bit though as it didn't apply cleanly to 1.3.5.

**Comment by Albrecht-S** [comment]
@CendioOssman Thank you for confirmation, it's very much appreciated. I'm glad it works for you.

> I had to massage the patch a bit though as it didn't apply cleanly to 1.3.5.

Yes, I noticed a difference of 41 lines of your line numbers WRT mine. The reason is obviously that you used the released 1.3.5 version and I used current git `branch-1.3`. Commit b846d2623386b782aec89dc17e944c65a484cd1f added "_support for copy+paste of image within one app_" which modified `src/Fl_x.cxx`.

I'll commit my patch ASAP...

**Comment by Albrecht-S** [comment]
OK, I pushed the commit 7ce6d2cf5dfc0488ec30d9f9f1709be73353479c to `branch-1.3` which should fix the issue for upcoming release 1.3.6 (no release date yet, but hopefully soon; working on it ...).

Still to do: commit / push the solution to `master`.

**Comment by Albrecht-S** [comment]
This is now fixed in master (1.4) as well (commit ff0543524b7b2df4931aaa4e81bdfa9fe82b76f0).
